<h3>Задача семинара</h3>

<p>Мы будем решать задачу, о которой говорили на лекции &mdash;&nbsp;реализуем&nbsp;механизм подсказок (<em>suggest</em>) категории по заголовку.</p>

<p>Представьте, что вы хотите разместить объявление о продаже шин. Чтобы положить своё объявление в нужную категорию, вам необходимо выбрать одну из 10 верхнеуровневых категорий, одну из 5 на следющем уровне, одну из 12 на следующем уровне, одну из 5 на последнем уровне.</p>

<p>Даже если варианты правильных&nbsp;категорий&nbsp;на каждом уровне очевидны, процесс подачи объявления лучше упростить.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="259" src="http://labelimages.avito.ru/mlcourse_week3_without_suggest.png" width="563" />
<figcaption>Пример поиска нужной категории без подсказок</figcaption>
</figure>
</div>

<p>&nbsp;Упрощённый процесс подачи объявлений может выглядеть следующим образом. Пользователь заполняет заголовок объявления, затем выбирает один из предложенных вариантов категорий, который больше всего подходит его объявлению.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="324" src="http://labelimages.avito.ru/mlcourse_week3_with_suggest.png" width="534" />
<figcaption>Работа suggest при публикации объявления</figcaption>
</figure>
</div>

<p>Задача &mdash; определить категорию объявления, используя заголовок. Всего в выборке будет 54 категории.</p>


<h3><strong>Многоклассовая классификация</strong></h3>

<p>Для решения задачи будем использовать модель логистической регрессии – линейная модель для бинарной классификации. Как её использовать для решения задачи классификации на 54 класса? Оказывается, любой бинарный классификатор, который способен оценивать меру принадлежности объектов к одному из 2 классов, можно использовать для многоклассовой классификации. </p>

<p>Мнгоклассовая классификация на основе бинарного классификатора может быть реализована с помощью одной из двух стратегий:</p>

<ul>
	<li>Один против всех (<em>One-vs-Rest</em> или <em>One-vs-All</em>);</li>
	<li>One-vs-One.</li>
</ul>

<p> </p>

<p><strong>One-vs-All </strong></p>

<p>В стратегии <em>One-vs-All</em> строятся <span class="math-tex">\(k\)</span> классификаторов, отделяющих каждый класс от остальных. Например, для 3 классов нужно обучить 3 классификатора:</p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="436" name="Снимок экрана 2021-08-03 в 19.26.56.png" src="https://ucarecdn.com/2abe5812-3f6e-42cf-97a4-a35a6de58c6d/" width="513">
<figcaption>Многоклассовая классфикация по стратегии OVR</figcaption>
</figure>
</div>

<p>Помимо эффективности построения (необходимы только <span class="math-tex">\(k\)</span> классификаторов), одним из преимуществ этого подхода является его интерпретируемость. Поскольку для каждого класса обучался свой классификатор, можно анализировать каждый классификатор в отдельности, чтобы понять, почему модель приняла такое решение.</p>

<p><em>Предсказания на тестовой выборке для OVR.</em></p>

<p>Для объектов из тестовой выборки предсказывается класс, получивший максимальную вероятность в результате предсказаний <span class="math-tex">\(k\)</span> классификаторов:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \ a_{k}(x),\)</span></p>

<p>где <span class="math-tex">\(a_k(x) -\)</span> предсказание модели бинарной классификации для класса <span class="math-tex">\(k\)</span>.  Выражение<em> <em><em><em><em><em><em><em><span class="math-tex">\(\argmax\)</span></em></em></em></em></em></em></em></em> означает выбор <span class="math-tex">\(k\)</span>, при котором значение <span class="math-tex">\(a_{k}(x)\)</span> максимально.</p>

<p><em>На практике обычно используют OVR стратегию или алгоритмы с нативной поддержкой многоклассовой классификации – нейронные сети или алгоритмы, которые мы будем изучать дальше в этом курсе (RandomForest, KNN).</em></p>

<p> </p>

<p><strong>One-vs-One (опционально)</strong></p>

<p>В стратегии <em>One-vs-One </em>обучается один классификатор для каждой пары классов. При этом каждый классификатор обучается только на объектах, относящихся к этим классам: </p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="551" name="Снимок экрана 2021-08-01 в 23.43.15.png" src="https://ucarecdn.com/b6a950d0-a787-49de-b6cd-9f2faadb91c3/" width="637">
<figcaption>Многклассовая классификация стратегией OVO</figcaption>
</figure>
</div>

<p>Таким образом, требуется построить квадратичное число классификаторов <span class="math-tex">\(\frac{k(k-1)}{2}\)</span>.</p>

<p>При больших <span class="math-tex">\(k\)</span> может потребоваться много времени на обучение. Однако этот метод может быть полезен для алгоритмов, которые плохо масштабируются с увеличением размера выборки, так как каждая отдельная задача обучения классификатора включает в себя только объекты, принадлежащие двум выбранным классам. </p>

<p><em>Предсказания на тестовой выборке для OVO.</em></p>

<p>Для объектов из тестовой выборки все <span class="math-tex">\(\frac{k(k-1)}{2}\)</span> классификаторов делают предсказания, а финальный ответ получается в результате «жёсткого» голосования:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \sum_{m \neq k} [a_{km}(x) = 1],\)</span></p>

<p>где  <span class="math-tex">\(a_{km}(x) -\)</span>классификатор с позитивным классом <span class="math-tex">\(k\)</span> и негативным классом <span class="math-tex">\(m\)</span>.  Жёстким оно называется потому, что никак не учитывается, насколько уверены были классификаторы в своих предсказаниях.</p>

<p>Если несколько классов для объекта получили одинаковое количество голосов или нужно определить не только класс, но и меру принадлежности, производится «мягкое» голосование: выигрывает тот класс в котором суммарно больше уверены классификаторы.</p>


## EDA

<p>Начнем работу с данными с импорта нужных библиотек:</p>

In [1]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from sklearn.feature_extraction.text import TfidfVectorizer

<p>Прочитаем <a href="https://stepik.org/media/attachments/lesson/537384/suggest_train.csv" rel="noopener noreferrer nofollow">данные</a>:</p>


In [2]:
data = pd.read_csv('suggest_train.csv', index_col=0)
data.head(3)

title  category_id  \
item_id                                                                   
335330            Раствор. Товарный бетон от производителя           15   
137480                                    Детская кроватка           16   
242690   Кровать -качалка с ящиком для вещей, два полож...           34   

                                                      name  
item_id                                                     
335330   Для дома и дачи|Ремонт и строительство|Стройма...  
137480   Для дома и дачи|Мебель и интерьер|Кровати, див...  
242690   Личные вещи|Товары для детей и игрушки|Детская...

In [3]:
data.shape[0]

20000

In [4]:
#data = data.drop_duplicates(subset=['title'])
#data.shape[0]

Видим, что данные упорядочены по категориям. Это может привести к ошибке при разбиении данных. Перемешаем данные:

In [5]:
data = data.sample(frac=1).reset_index(drop=True)

Проверим, правильно ли закодированы категории:

In [6]:
np.sort(data.category_id.unique())

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53], dtype=int64)

In [7]:
data.category_id.min(), data.category_id.max()

(0, 53)

Категории закодированы правильно.

**Баланс классов**

Посмотрим на распределение классов. Большой дисбаланс может вызвать проблемы модели. Например, в топ-3 по вероятности будут попадать нерелевантные категории просто потому, что в датасете их намного больше, чем остальных.

In [50]:
data.groupby('category_id').size().describe()

count     54.000000
mean     370.370370
std       33.178361
min      309.000000
25%      346.000000
50%      361.000000
75%      398.750000
max      447.000000
dtype: float64

Видим, что минимальное количество примеров на категорию 309, а максимальное 447. Скорее всего, проблемы не возникнет.

**Примеры заголовков**

Посмотрим на примеры топ-3 наиболее и наименее популярных категорий:

In [300]:
data[data.name.isin(data.name.value_counts()[:3].index)].drop_duplicates(subset='name')

title  category_id  \
12                                   Повязка на руку           37   
67  Наволочки простыни комплекты для дома и гостиниц           27   
69                                    Продам трильяж           23   

                                                 name  
12  Личные вещи|Красота и здоровье|Приборы и аксес...  
67  Для дома и дачи|Мебель и интерьер|Текстиль и к...  
69           Для дома и дачи|Мебель и интерьер|Другое

Наименее популярные:

In [301]:
data[data.name.isin(data.name.value_counts()[-3:].index)].drop_duplicates(subset='name')

title  category_id  \
9                      Кольцо золото           40   
52   Телевизор Panasonic TX-PR65VT20            5   
382  Apple MacBook Air 13 early 2015            1   

                                                  name  
9       Личные вещи|Часы и украшения|Ювелирные изделия  
52   Бытовая электроника|Аудио и видео|Телевизоры и...  
382                       Бытовая электроника|Ноутбуки

Видим, что в бытовой технике могут быть довольно редкие сочетания букв, например "K50AB", есть слова с орфографическими ошибками ("гаршок"). Есть и частые слова, которые встретились в 2 из 6 объявлений. 

**Названия категорий**

Посмотрим на названия категорий:

In [302]:
data.name.unique()[:3]

array(['Хобби и отдых|Спорт и отдых|Другое',
       'Для дома и дачи|Мебель и интерьер|Освещение',
       'Бытовая электроника|Товары для компьютера|Сетевое оборудование'],
      dtype=object)

Видим, что категории могут находиться на разном уровне иерархии. Посмотрим, насколько сильно может отличаться глубина категории:

In [303]:
data.name.apply(lambda x: len(x.split('|'))).value_counts().sort_index()

2     1406
3    15973
4     2621
Name: name, dtype: int64

Можно подумать, как использовать данные о том, что у классов есть иерархическая структура.

**Статистики по текстам**

Посмотрим, отличаются ли статистически тексты заголовков для разных категорий. Посчитаем количество уникальных символов в заголовке:

In [304]:
data.title.apply(len).describe()

count    20000.000000
mean        25.093300
std         12.474148
min          5.000000
25%         15.000000
50%         23.000000
75%         34.000000
max         50.000000
Name: title, dtype: float64

Посмотрим на на распределение количества слов:

In [305]:
data.title.apply(lambda x: len(x.split())).describe()

count    20000.00000
mean         3.79825
std          1.96368
min          1.00000
25%          2.00000
50%          3.00000
75%          5.00000
max         14.00000
Name: title, dtype: float64

В среднем в заголовках 3-4 слова. Заголовки короткие

Интересно посмотреть на краевые случаи. Давайте выведем примеры, где в заголовок получилось уместить больше 12 слов:

In [306]:
data[data.title.apply(lambda x: len(x.split()) > 12)].head(3)

title  category_id  \
3357     HP 17.3 / Core i5 / 4 озу / 2 видео/ как новый            1   
4272    Отдам 2 альбома для фото 10 х 15 на 96 и 48 фот           19   
8873  PS3 Slim + Ps Move + Ps Eye + 3 джойстика + 12...            4   

                                                   name  
3357                       Бытовая электроника|Ноутбуки  
4272  Для дома и дачи|Мебель и интерьер|Предметы инт...  
8873  Бытовая электроника|Игры, приставки и программ...

Все примеры из бытовой электроники. Значит, количество слов может послужить предиктором категории. Посчитаем статистики для категорий:

In [307]:
data['word_count'] = data.title.apply(lambda x: len(x.split()))
data.groupby('name')['word_count'].mean().sort_values().head(3)

name
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Платья и юбки     2.576190
Для дома и дачи|Мебель и интерьер|Шкафы и комоды                       2.756219
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Верхняя одежда    2.824074
Name: word_count, dtype: float64

Действительно, среднее количество слов для категории отличается в 2 раза.

In [308]:
data.groupby('name')['word_count'].mean().sort_values().tail(3)

name
Хобби и отдых|Велосипеды|Запчасти и аксессуары            5.508600
Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки    5.531429
Хобби и отдых|Билеты и путешествия|Путешествия            5.641399
Name: word_count, dtype: float64

Кажется, что статистики по тексту могут помочь улучшить качество работы классификатора. С другой стороны, нужно хорошенько подумать, стоит ли использовать такие признаки в сервисе?  

## Бейзлайн

Разобьем данные на обучающую выборку и валидацию:

In [6]:
train_titles, val_titles, y_train, y_val = train_test_split(
    data['title'],
    data['category_id'],
    random_state=42, stratify=data['category_id'], test_size=0.05
)

In [373]:
train_titles

3639                  Продаю входную металлическую дверь
2489                                              Стенка
19136                                     Системный блок
4040       Наволочка двухцветкая 50x70 Бельпостель новая
3257                     Лампа светодиодная для растений
                              ...                       
19076                             Платья желтое барадина
19568                     iPhone 4 в идеальном состоянии
4089                   Моноблок 18.5" Acer Aspire Z1-601
17823                 Детская кровать "Лютик" с матрасом
19400    Виброплита поступательного движения grost VH 80
Name: title, Length: 19000, dtype: object

Обычно, в задачах классификации первым бейзлайном является предсказание самым популярным классом. Но когда выборка сбалансирована, а классов много, такой бейзлайн не очень информативен:

In [6]:
(y_val == y_train.mode().values[0]).mean()

0.022

## CountVectorizer 

Для получения матриц признаков преобразуем заголовки с помощью CountVectorizer():

In [33]:
count_vectorizer = CountVectorizer()

X_train = count_vectorizer.fit_transform(train_titles.values)
X_val = count_vectorizer.transform(val_titles.values)

In [8]:
count_vectorizer = TfidfVectorizer(analyzer='char_wb', ngram_range=(3, 6), min_df=1)
X_train = count_vectorizer.fit_transform(train_titles.values)
X_val = count_vectorizer.transform(val_titles.values)

In [9]:
X_train

<19000x217877 sparse matrix of type '<class 'numpy.float64'>'
	with 1272231 stored elements in Compressed Sparse Row format>

<p>В данном случае вместо&nbsp;<span class="math-tex">\(15000 \times 17316=259740000\)</span>&nbsp;значений мы храним всего&nbsp;<span class="math-tex">\(54676\)</span> тройки&nbsp;<span class="math-tex">\((i, j, x_j^{(i)})\)</span>. Это&nbsp;1500&nbsp;раз меньше!</p>

<p>Sparse формат не доставляет никаких неудобств &mdash;&nbsp;все полезные функции можно применять к матрицам в разреженном формате. Единственное &laquo;неудобство&raquo;&nbsp;&mdash;&nbsp;нельзя посмотреть&nbsp;на матрицу.<span style="color:#a03881"> </span>Но в данном случае этого делать не нужно, потому что всё, что мы увидим, это матрица&nbsp;из нулей. Не повторяйте этого на практике, на больших&nbsp;выборках преобразованные данные могут не поместиться в оперативную память:</p>

In [35]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

Посмотрим на получившийся словарь:

In [36]:
print('words in vocab: ', len(count_vectorizer.vocabulary_))
count_vectorizer.vocabulary_

words in vocab:  20203


{'шторы': 19828,
 'тюль': 18667,
 'тарелки': 18229,
 'глубокие': 10593,
 'продам': 16201,
 'откосы': 15051,
 'обналичку': 14747,
 'счетчик': 18126,
 'воды': 10171,
 '15': 420,
 'бетар': 9425,
 'kop': 5330,
 'импорт': 11896,
 'исландия': 11996,
 '10': 120,
 'эйре': 19890,
 '1961': 699,
 'фанки': 18927,
 'тайм': 18179,
 'композ': 12644,
 'n12': 6105,
 'samsung': 7164,
 'galaxy': 4546,
 's3': 7086,
 'duos': 3916,
 'шарм': 19620,
 'sunlight': 7624,
 'крышка': 13034,
 'zepter': 8646,
 'synchro': 7671,
 'click': 3436,
 'панельку': 15197,
 'на': 14313,
 'iphone': 5095,
 '5s': 1860,
 'grand': 4670,
 'theft': 7797,
 'auto': 2783,
 'ps': 6715,
 'телевизор': 18277,
 'lg': 5490,
 'сапоги': 17093,
 'фрг': 19134,
 'марок': 13668,
 '1990': 731,
 'серебро': 17299,
 '800': 2198,
 'лет': 13298,
 'тевтонскому': 18265,
 'орде': 14943,
 'пуховая': 16381,
 'косынка': 12839,
 'sony': 7457,
 'xperia': 8540,
 'z2': 8602,
 'андорра': 8982,
 'из': 11831,
 'москвы': 14172,
 'часы': 19473,
 'apple': 2662,
 'watch'

Убедимся, что необходимые значения хранятся по индексам из словаря:

In [37]:
count_vectorizer.transform(['Продам продам'])[0, count_vectorizer.vocabulary_['продам']]

2

Видим, что CountVectorizer автоматически приводит слова к нижнему регистру.

Важные параметры

Класс CountVectorizer содержит все нужные параметры для настройки преобразования под вашу задачу:

<pre>
<code class="language-python">CountVectorizer(
    *,
    input='content',
    encoding='utf-8',
    decode_error='strict',
    strip_accents=None,
    lowercase=True,
    preprocessor=None,
    tokenizer=None,
    stop_words=None,
    token_pattern='(?u)\\b\\w\\w+\\b',
    ngram_range=(1, 1),
    analyzer='word',
    max_df=1.0,
    min_df=1,
    max_features=None,
    vocabulary=None,
    binary=False,
    dtype=&lt;class 'numpy.int64'&gt;,
)</code></pre>

<p>Рассмотрим наиболее важные из них:</p>

<ul>
	<li><code>max_df</code>, <code>min_df</code>&nbsp;&mdash;&nbsp;параметры, позволяющие фильтровать слова по порогу встречаемости. Если вы передаёте в параметры действительные числа от 0 до 1, это это максимальная и минимальная доля встречаемости слова. Если целые, то абсолютное значение встречаемости;</li>
	<li><code>ngram_range</code>&nbsp;&mdash;&nbsp;диапазон n-грамм: (1, 3) &mdash;&nbsp;значит, включить в словарь&nbsp;униграммы, биграммы и тригаммы;</li>
	<li><code>analyzer</code>&nbsp;&mdash;&nbsp;определяет, что считать токеном. Есть возможность токенезировать по словам, буквам и буквам с учётом окончаний слов. Может работать особенно эффективно в сочетании с учётом n-грамм;</li>
	<li><code>token_pattern</code>&nbsp;&mdash; регулярное выражение для токенизации.</li>
</ul>


## Модель логистической регрессии

Обучим модель логистической регрессии c параметрами по умолчанию и посчитаем accuracy:

In [10]:
model = LogisticRegression()
model.fit(X_train, y_train)

preds = model.predict(X_val)
accuracy_score(preds, y_val)

0.781

Логистическая регрессия позволяет получить не только предсказанные категории, но и вероятности. Предсказания для одного объекта в нашем случае выглядят как вектор из 54 чисел:

In [ ]:
np.round(model.predict_proba(X_val[0]), 3)

Проверим свою интуицию по поводу весов полученной модели. Мы знаем, что слова встречающиеся в разных категориях, не должны получить большой вес в модели. Посмотрим на вес слова 'продам' для разных категорий:

In [595]:
model.coef_[:, count_vectorizer.vocabulary_['продам']]

KeyError: 'продам'

Слова, по которым можно определить категорию, наоборот, должны получить большой вес. Посмотрим на вес для слова 'iphone' для разных категорий:

In [15]:
model.coef_[:, count_vectorizer.vocabulary_['iphone']]

array([10.56217783, -0.28302724, -0.25696417, -0.46325024, -0.26310061,
       -0.22940678,  1.11852941, -0.27567361, -0.34411253,  4.99828178,
       -0.24306577, -0.16299175, -0.35636681, -0.22306383, -0.31211045,
       -0.48431989, -0.21636626, -0.44171829, -0.27689242, -0.39043365,
       -0.40364638, -0.42376902, -0.25637711, -0.48936954, -0.40933282,
       -0.20462596, -0.25986752, -0.43872495, -0.31017565, -0.39299887,
       -0.42607865, -0.05391445, -0.25880853, -0.24327095, -0.28220237,
       -0.20141102, -0.25101834, -0.51647912, -0.36966218, -0.28072492,
       -0.28712178, -0.29960978, -0.26893679, -0.41788875, -0.43457001,
       -0.34924486, -0.09181143, -0.48711239, -0.33148956, -0.4553737 ,
       -0.32274278, -0.34724939, -0.40069912, -0.48981603])

Действительно, вес при этом слове для некоторых категорий очень большой. Ведь если это слово есть в заголовке, скорее всего, правильной категорией будет что-то начинающееся на 'Бытовая электроника|Телефоны|'. Проверим, так ли это:

In [16]:
cat_names = data.set_index('category_id')['name'].to_dict()
top_iphone_cats = np.argsort(model.coef_[:, count_vectorizer.vocabulary_['iphone']])[-3:]
[cat_names[c] for c in top_iphone_cats]

['Бытовая электроника|Телефоны|Другие марки',
 'Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки',
 'Бытовая электроника|Телефоны|iPhone']

In [17]:
np.argsort(model.coef_[:, count_vectorizer.vocabulary_['iphone']])[-3:]

array([6, 9, 0], dtype=int64)

In [18]:
data.set_index('category_id')['name'][9]

category_id
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
                           ...                        
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
9    Бытовая электроника|Телефоны|Аксессуары|Чехлы ...
Name: name, Length: 350, dtype: object

**Ошибки модели**

Посмотрим, в каких категориях чаще всего происходят ошибки. Для этого создадим удобный dataframe:

In [19]:
models_erros = pd.DataFrame({
    'title': val_titles.values,
    'preds_category': [cat_names[cat] for cat in preds],
    'true_category': [cat_names[cat] for cat in y_val],
    'preds_ok': preds == y_val
})

Выведем 3 случайные ошибки модели:

In [20]:
models_erros[~models_erros['preds_ok']].sample(3)

title  \
item_id                                             
130675                             Leagoo elite 1   
118153   Продаю сумку для детских принадлежностей   
5677                               Yamaha RBX 170   

                                            preds_category  \
item_id                                                      
130675      Хобби и отдых|Велосипеды|Запчасти и аксессуары   
118153    Личные вещи|Одежда, обувь, аксессуары|Аксессуары   
5677     Хобби и отдых|Музыкальные инструменты|Для студ...   

                                             true_category  preds_ok  
item_id                                                               
130675           Бытовая электроника|Телефоны|Другие марки     False  
118153   Личные вещи|Товары для детей и игрушки|Товары ...     False  
5677     Хобби и отдых|Музыкальные инструменты|Гитары и...     False

Видим, что в первых двух строчках ошибки, скорее всего, произошли по причине опечаток. Нужно подумать над тем, как повысить устойчивость модели к опечаткам и ошибкам. Этим вам предстоит заняться в домашнем задании.

На этом анализ ошибок, конечно же, заканчивать не стоит — но мы не будем продолжать это делать здесь, а оставим в виде упражнения.

## Word2vec

Импортируем gensim – библиотеку, с помощью которой будем обучать модель w2v. Также создадим понадобятся вспомогательные классы, которые будут логировать ошибку после каждой итерации обучения.

In [328]:
pip install gensim

Note: you may need to restart the kernel to use updated packages.


In [39]:
import re
import gensim
from gensim.models.word2vec import Word2Vec
from gensim.models.fasttext import FastText
from gensim.models.callbacks import CallbackAny2Vec

class LossLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        if self.epoch == 0:
            print('Loss after epoch {}: {}'.format(self.epoch, loss))
        else:
            print('Loss after epoch {}: {}'.format(self.epoch, loss - self.loss_previous_step))
        self.epoch += 1
        self.loss_previous_step = loss
        
class EpochLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0

def on_epoch_end(self, model):
    print(f'Epoch {self.epoch}')
    self.epoch += 1

print(gensim.__version__)

4.2.0


Прочитаем данные и выведем первые несколько строк:

In [374]:
w2v_train_data = pd.read_csv('unlabeled_data.csv',  index_col=0)
print(w2v_train_data.shape)
w2v_train_data.head()

(200000, 2)


item_id                                              title
30000   377939                            1 рубль царские 10 штук
30001   384779  Напольный насос giyo высокого давления, с мано...
30002   276889                                Sony Xperia m2 Dual
30003   155718   Zoom B2.1U+ AC Процессор эффектов для бас-гитары
30004   260705                    Люстра Favourite petite 1574-7P

Данные для обучения word2vec — это просто тексты или любые другие наборы последовательностей токенов, для которых можно предположить справедливость дистрибутивной гипотезы. Токенизируем тексты:

In [375]:
WORD_PATTERN = '(?u)\\b\\w\\w+\\b'  # паттерн для токенизации возьмем из векторайзеров sklearn
reg_exp = re.compile(pattern=WORD_PATTERN)  # скомпилируем регулярное выражение
sentences = [reg_exp.findall(s.lower()) for s in w2v_train_data.title]

Выведем результат токенизации:

In [376]:
sentences[:3]

[['рубль', 'царские', '10', 'штук'],
 ['напольный', 'насос', 'giyo', 'высокого', 'давления', 'манометр'],
 ['sony', 'xperia', 'm2', 'dual']]

In [377]:
1e-6

1e-06

Обучим w2v с параметрами по усмолчанию: 

In [502]:
w2v_model = Word2Vec(vector_size=220, window=10, min_count=2, sg=1, negative=10, ns_exponent=0.25, shrink_windows=True)
w2v_model.build_vocab(sentences)
w2v_model.train(
   corpus_iterable=sentences,
   total_examples=w2v_model.corpus_count,
   epochs=40,
   compute_loss=True,
   callbacks=[LossLogger()]
)

Loss after epoch 0: 2826708.25
Loss after epoch 1: 1292064.25
Loss after epoch 2: 923431.0
Loss after epoch 3: 799357.5
Loss after epoch 4: 716546.0
Loss after epoch 5: 678896.0
Loss after epoch 6: 628815.5
Loss after epoch 7: 585835.5
Loss after epoch 8: 506691.0
Loss after epoch 9: 479640.0
Loss after epoch 10: 452869.0
Loss after epoch 11: 432606.0
Loss after epoch 12: 409773.0
Loss after epoch 13: 392829.0
Loss after epoch 14: 373970.0
Loss after epoch 15: 359233.0
Loss after epoch 16: 344516.0
Loss after epoch 17: 332241.0
Loss after epoch 18: 320822.0
Loss after epoch 19: 310278.0
Loss after epoch 20: 302189.0
Loss after epoch 21: 292632.0
Loss after epoch 22: 285791.0
Loss after epoch 23: 278602.0
Loss after epoch 24: 272534.0
Loss after epoch 25: 267272.0
Loss after epoch 26: 258635.0
Loss after epoch 27: 254973.0
Loss after epoch 28: 251336.0
Loss after epoch 29: 246489.0
Loss after epoch 30: 242534.0
Loss after epoch 31: 242297.0
Loss after epoch 32: 235883.0
Loss after epoch

(25552454, 29369480)

Попытаемся понять, получилось ли обучиться. Выведем векторы, близкие к вектору слова 'продам';

In [503]:
w2v_model.wv.similar_by_word('продам')

[('медицинскую', 0.4443364143371582),
 ('новую', 0.43903228640556335),
 ('хорошую', 0.42489495873451233),
 ('советскую', 0.4214405417442322),
 ('прдаю', 0.4194292724132538),
 ('напольную', 0.4185582399368286),
 ('продадим', 0.4166226387023926),
 ('кожаную', 0.41509345173835754),
 ('красивую', 0.4127601385116577),
 ('боксерскую', 0.40963295102119446)]

Посмотрим на ещё одно слово. Интересно, что можно поискать слова близкие к слову 'айфон' и тем самым получить написание этого слова на английском:

In [504]:
w2v_model.wv.similar_by_word('айфон')

[('lunatic', 0.7184357047080994),
 ('4s8', 0.70102858543396),
 ('taktik', 0.671530544757843),
 ('5с', 0.6523693203926086),
 ('32гиг', 0.647223949432373),
 ('броне', 0.6435747742652893),
 ('айфон4', 0.6407101154327393),
 ('3gs', 0.6404737830162048),
 ('ronin', 0.6320513486862183),
 ('покупала', 0.6306703090667725)]

Напишем класс, которые позволит там получать векторы для заголовков, как сумму векторов входящих в него слов:

In [505]:
class Word2VecTransformer:
    
    def __init__(self, w2v_model, word_pattern):
        
        self.w2v_model = w2v_model
        self.word_pattern = word_pattern
        self.re = re.compile(pattern=self.word_pattern)
        
    def fit(self, X):
        return self
    
    def transform(self, X):
        
        X_transformed = np.zeros((len(X), self.w2v_model.wv.vector_size))
        for i, title in enumerate(X):
            
            k = 1
            title_vector = np.zeros((self.w2v_model.wv.vector_size,))
            tokens = self.re.findall(title.lower())
            for token in tokens:
                if token in self.w2v_model.wv.key_to_index:
                    title_vector += self.w2v_model.wv.get_vector(token)
                    k += 1
                    
            X_transformed[i] = title_vector / k
        
        return X_transformed

Трансформируем с помощью него обучающую выборку и обучим логистическую регрессию:

In [506]:
w2v_transformer = Word2VecTransformer(w2v_model=w2v_model, word_pattern=WORD_PATTERN)

train_w2v = w2v_transformer.transform(train_titles.values)
val_w2v = w2v_transformer.transform(val_titles.values)

model_on_w2v = LogisticRegression(solver='saga')
model_on_w2v.fit(train_w2v, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='saga', tol=0.0001, verbose=0,
                   warm_start=False)

In [507]:
# model_on_w2v = LogisticRegression(solver='liblinear')
# model_on_w2v.fit(train_w2v, y_train)

In [508]:
# from sklearn.linear_model import SGDClassifier
# model_on_w2v = SGDClassifier(loss='log')
# model_on_w2v.fit(train_w2v, y_train)

Оценим качество классификатора:

In [509]:
preds = model_on_w2v.predict(val_w2v)
accuracy_score(preds, y_val)

0.793

In [510]:
result = np.sort(model_on_w2v.predict_proba(val_w2v))[:, -3:]
np.argsort(model_on_w2v.predict_proba(val_w2v))[:, -3:]

array([[19, 23, 22],
       [24, 23, 27],
       [19, 23, 36],
       ...,
       [14,  6,  3],
       [ 6,  9,  0],
       [15, 23, 25]], dtype=int64)

In [511]:
def get_acc(preds, y_answers):
    counter = 0
    for i, y in enumerate(y_answers):
        if y in preds[i]:
            counter += 1
    return counter / len(preds)
        

In [512]:
get_acc(np.argsort(model_on_w2v.predict_proba(val_w2v))[:, -3:], y_val)

0.921

Варианты улучшений: 

* Word2vec не может получить вектора слов, которых не было при обучении модели. Однако есть модель, очень похожая на word2vec, но использующая при обучении буквенные n-гаммы. Эта модель называется Fasttext — в библиотеке gensim есть её реализация;
* Метод агрегации, который мы использовали, очень грубый: можно попробовать не суммировать, а считать среднее или суммировать с весами;
* Можно попробовать подобрать более оптимальные гиперпараметры w2v: использовать skipgram вместо CBOW, учить с другим размером окна, использовать другой размер вектора, изменить соотношение позитивных и негативных примеров в negative sampling или использовать иерархический софтмакс.

# Для STEPIK

In [513]:
df = pd.read_csv('suggest_test.csv')
df.head()

title
0                               Телевизор Samsung
1        Игрушки ручной работы (вязание, валяние)
2                  Раковина с ножкой и смесителем
3                                  Куколки mattel
4  Автомобильный сабвуфер активный Kicx ICQ-300BA

In [514]:
test_w2v = w2v_transformer.transform(df.title)

In [515]:
len(test_w2v)

10000

In [516]:
result = np.argsort(model_on_w2v.predict_proba(test_w2v))[:, -3:]
np.argsort(model_on_w2v.predict_proba(test_w2v))[:, -3:]

array([[13,  2,  5],
       [44, 27, 30],
       [19, 23, 21],
       ...,
       [30, 41, 32],
       [21, 30, 26],
       [23, 24, 15]], dtype=int64)

In [517]:
pd.DataFrame(result).to_csv('solution.csv', header=['top1', 'top2', 'top3'], index=False)

In [416]:
ft_model = FastText(vector_size=50, window=5, min_count=2, negative=15, min_n=2, max_n=4)
ft_model.build_vocab(sentences)

In [417]:
ft_model.train(
   corpus_iterable=sentences,
   total_examples=ft_model.corpus_count,
   epochs=30,
   compute_loss=True,
   callbacks=[LossLogger()]
)

Loss after epoch 0: 0.0
Loss after epoch 1: 0.0
Loss after epoch 2: 0.0
Loss after epoch 3: 0.0
Loss after epoch 4: 0.0
Loss after epoch 5: 0.0
Loss after epoch 6: 0.0
Loss after epoch 7: 0.0
Loss after epoch 8: 0.0
Loss after epoch 9: 0.0
Loss after epoch 10: 0.0
Loss after epoch 11: 0.0
Loss after epoch 12: 0.0
Loss after epoch 13: 0.0
Loss after epoch 14: 0.0
Loss after epoch 15: 0.0
Loss after epoch 16: 0.0
Loss after epoch 17: 0.0
Loss after epoch 18: 0.0
Loss after epoch 19: 0.0
Loss after epoch 20: 0.0
Loss after epoch 21: 0.0
Loss after epoch 22: 0.0
Loss after epoch 23: 0.0
Loss after epoch 24: 0.0
Loss after epoch 25: 0.0
Loss after epoch 26: 0.0
Loss after epoch 27: 0.0
Loss after epoch 28: 0.0
Loss after epoch 29: 0.0


(19163645, 22027110)

In [404]:
ft_model.wv.similar_by_word('продам')

[('продаю', 0.989808976650238),
 ('продаём', 0.9528104662895203),
 ('продаи', 0.948464572429657),
 ('продадим', 0.9368908405303955),
 ('продаёться', 0.9241591095924377),
 ('продаем', 0.9204390048980713),
 ('продаеться', 0.9085977077484131),
 ('продать', 0.8988447189331055),
 ('проодам', 0.8977782130241394),
 ('продажу', 0.8927575349807739)]

In [405]:
class FastTextTransformer:
    
    def __init__(self, ft_model, word_pattern):
        
        self.ft_model = ft_model
        self.word_pattern = word_pattern
        self.re = re.compile(pattern=self.word_pattern)
        
    def fit(self, X):
        return self
    
    def transform(self, X):
        
        X_transformed = np.zeros((len(X), self.ft_model.wv.vector_size))
        for i, title in enumerate(X):
            
            title_vector = np.zeros((self.ft_model.wv.vector_size,))
            tokens = self.re.findall(title.lower())
            k = 1
            for token in tokens:
                if token in self.ft_model.wv.key_to_index:
                    title_vector += self.ft_model.wv.get_vector(token)
                    k += 1
                    
            X_transformed[i] = title_vector / k
        
        return X_transformed

In [418]:
ft_transformer = FastTextTransformer(ft_model=ft_model, word_pattern=WORD_PATTERN)

train_ft = ft_transformer.transform(train_titles.values)
val_ft = ft_transformer.transform(val_titles.values)

model_on_ft = LogisticRegression(solver='liblinear')
model_on_ft.fit(train_ft, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [419]:
preds = model_on_ft.predict(val_ft)
accuracy_score(preds, y_val)

0.6822

In [420]:
get_acc(np.argsort(model_on_ft.predict_proba(val_ft))[:, -3:], y_val)

0.8256

In [421]:
ft_model.corpus_count

200000

In [382]:
w2v_model.corpus_count

200000